portal/flatpak-portal.c

```diff
@@ -700,6 +741,479 @@ authorize_method_handler (GDBusInterfaceSkeleton *interface,
+                                          m->obj_path,
+                                          "org.freedesktop.portal.Flatpak",
+                                          "UpdateAvailable",
+                                          g_variant_new ("(a{sv})", &builder),
+                                          &error))
+        {
+          g_warning ("Failed to emit UpdateAvailable: %s", error->message);
+        }
+    }
+}
+
+static void
+check_all_for_updates_in_thread_func (GTask *task,
+                                      gpointer source_object,
+                                      gpointer task_data,
+                                      GCancellable *cancellable)
+{
+  GList *monitors, *l;
+
+  monitors = update_monitors_get_all (NULL);
+
+  for (l = monitors; l != NULL; l = l->next)
+    {
+      PortalFlatpakUpdateMonitor *monitor = l->data;
+      UpdateMonitorData *m = update_monitor_get_data (monitor);
+      gboolean was_closed = FALSE;
+
+      g_mutex_lock (&m->lock);
+      if (m->closed)
+        was_closed = TRUE;
+      else
+        m->running = TRUE;
+      g_mutex_unlock (&m->lock);
+
+      if (!was_closed)
+        {
+          check_for_updates (monitor);
+
+          g_mutex_lock (&m->lock);
+          m->running = FALSE;
+          if (m->closed) /* Was closed during running, do delayed close */
+            update_monitor_do_close (monitor);
+          g_mutex_unlock (&m->lock);
+        }
+    }
+
+  g_list_free_full (monitors, g_object_unref);
+
+  G_LOCK (update_monitors);
+  update_monitors_timeout_running_thread = FALSE;
+
+  if (g_hash_table_size (update_monitors) > 0)
+    update_monitors_timeout = g_timeout_add_seconds (UPDATE_POLL_TIMEOUT_SEC, check_all_for_updates_cb, NULL);
+
+  G_UNLOCK (update_monitors);
+}
+
+/* Runs on main thread */
+static gboolean
+check_all_for_updates_cb (void *data)
+{
+  g_autoptr(GTask) task = g_task_new (NULL, NULL, NULL, NULL);
+
+  g_debug ("Checking all update monitors");
+
+  G_LOCK (update_monitors);
+  update_monitors_timeout = 0;
+  update_monitors_timeout_running_thread = TRUE;
+  G_UNLOCK (update_monitors);
+
+  g_task_run_in_thread (task, check_all_for_updates_in_thread_func);
+
+  return G_SOURCE_REMOVE; /* This will be re-added by the thread when done */
+}
+
+/* Runs in worker thread */
+static gboolean
+handle_create_update_monitor (PortalFlatpak *object,
+                              GDBusMethodInvocation *invocation,
+                              GVariant *options)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(PortalFlatpakUpdateMonitorSkeleton) monitor = NULL;
+  const char *sender;
+  g_autofree char *sender_escaped = NULL;
+  g_autofree char *obj_path = NULL;
+  g_autofree char *token = NULL;
+  g_autoptr(GError) error = NULL;
+  int i;
+
+  if (!g_variant_lookup (options, "token", "s", &token))
+    token = g_strdup_printf ("%d", g_random_int_range (0, 1000));
```

Comment:
Well, its scoped by the sender id, so it will only happen if the same process opens multiple monitors for itself without specifying a token, so it doesn't really seem like a problem.
What would happen is that the g_dbus_interface_skeleton_export() call fails and we'll return an error.